mr/src/main/java/org/apache/iceberg/mr/Catalogs.java

```diff
@@ -180,47 +186,80 @@ public static boolean dropTable(Configuration conf, Properties props) {
-      CatalogLoader loader = (CatalogLoader) DynConstructors.builder(CatalogLoader.class)
-              .impl(catalogLoaderClass)
-              .build()
-              .newInstance();
-      Catalog catalog = loader.load(conf);
-      LOG.info("Loaded catalog {} using {}", catalog, catalogLoaderClass);
-      return Optional.of(catalog);
+  static Optional<Catalog> loadCatalog(Configuration conf, String catalogName) {
+    String catalogType;
+    String name = catalogName;
+    if (name == null) {
+      name = "default";
     }
+    catalogType = conf.get(String.format(InputFormatConfig.CATALOG_TYPE_TEMPLATE, name));
 
-    String catalogName = conf.get(InputFormatConfig.CATALOG);
+    // keep both catalog configuration methods for seamless transition
+    if (catalogType != null) {
+    // new logic
+      return loadCatalog(conf, catalogType, String.format(InputFormatConfig.CATALOG_WAREHOUSE_TEMPLATE, name),
+              String.format(InputFormatConfig.CATALOG_LOADER_CLASS_TEMPLATE, name));
+    } else {
+      // old logic
+      // use catalog {@link InputFormatConfig.CATALOG} stored in global hive config if table specific catalog
+      // configuration or default catalog definition is missing
+      catalogType = conf.get(InputFormatConfig.CATALOG);
+      return loadCatalog(conf, catalogType, InputFormatConfig.HADOOP_CATALOG_WAREHOUSE_LOCATION,
+              InputFormatConfig.CATALOG_LOADER_CLASS);
+    }
+  }
 
-    if (catalogName != null) {
-      Catalog catalog;
-      switch (catalogName.toLowerCase()) {
-        case HADOOP:
-          String warehouseLocation = conf.get(InputFormatConfig.HADOOP_CATALOG_WAREHOUSE_LOCATION);
+  private static Optional<Catalog> loadCatalog(Configuration conf, String catalogType,
+                                               String warehouseLocationConfigName, String loaderClassConfigName) {
+    if (catalogType == null) {
+      LOG.info("Catalog is not configured");
+      return Optional.empty();
+    }
 
-          catalog = (warehouseLocation != null) ? new HadoopCatalog(conf, warehouseLocation) : new HadoopCatalog(conf);
-          LOG.info("Loaded Hadoop catalog {}", catalog);
+    Catalog catalog;
+    switch (catalogType.toLowerCase()) {
+      case HADOOP:
+        String warehouseLocation = conf.get(warehouseLocationConfigName);
+        catalog = (warehouseLocation != null) ? new HadoopCatalog(conf, warehouseLocation) :
+                new HadoopCatalog(conf);
+        LOG.info("Loaded Hadoop catalog {}", catalog);
+        return Optional.of(catalog);
+      case HIVE:
+        catalog = HiveCatalogs.loadCatalog(conf);
+        LOG.info("Loaded Hive Metastore catalog {}", catalog);
+        return Optional.of(catalog);
+      case CUSTOM:
+        String catalogLoaderClass = conf.get(loaderClassConfigName);
+        if (catalogLoaderClass != null) {
+          CatalogLoader loader = (CatalogLoader) DynConstructors.builder(CatalogLoader.class)
```

Comment:
I wonder if the `CatalogLoader` class is needed at all. `CatalogUtil` already has a method to get a catalog from class name and configuration parameters which is used in Spark and Flink.

The `CatalogLoader` interface also creates potentially weird dependencies between the catalog modules and the mr module. When implementing this interface for the Nessie catalog I had to depend on the mr module for only that interface. 

I would be in favour of Deprecating `CatalogLoader` in favour of `CatalogUtil`.

Reply:
I think formally we would deprecate it for the 0.12.0 release and remove it in 0.13.0. Given the relative newness of the Hive code it might be ok to just remove it though. I don't mind either way and will defer to the consensus on the mailing list.

Reply:
From my reading the 'old' logic and the 'new' logic for building a catalog seem to be interleved. I wonder if its possible to separate out the two paths.  There might be a bit of code duplication but I think it would be more readable and it'll be easier to remove the old path later.

Reply:
The logic is a lot clearer for me now. There is still a mix of constructor based, CatalogUtil and HiveCatalogs based catalog building. Feels to me like it could be collapsed even further into a CatalogUtil call

Reply:
HiveCatalogs handles the caching, but in the end, calls CatalogUtil.

Reply:
Agreed, I am looking for ways to simplify the logic in this method: if possible fewer branching statements and 1 way of instantiating a the resulting catalog.

Reply:
We've kept the hive code marked "experimental" because we knew that the catalog and table loading would change. I think it is nice to have some support for the old method of configuring tables, but we don't need to work too hard to continue to support it.